File: vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

#define ALPHABET_SIZE 26

int main(int argc, string argv[])
{
    //Check we have the right number of arguments
    if (argc != 2)
    {
        printf("Error: you must provide the key as a command line argument.\n");
        return 1;
    }


    //declare int key array, equal to the length of the string key entered by the user
    int key_len = strlen(argv[1]);
    int key[key_len];

    //Convert string key into array of ints
    for (int i = 0; i < key_len; i++)
    {
        //check i'th char of string key is alpha
        if (isalpha(argv[1][i]))
        {
            key[i] = toupper(argv[1][i]) - 'A';
        }
        else
        {
            printf("Error: key must be alphanumeric\n");
            return 1;
        }
    }

    printf("plaintext:");
    string plaintext = get_string();

    printf("ciphertext:");

    //Iterate through plaintext string
    for (int i = 0, j = 0, text_len = strlen(plaintext); i < text_len; i++)
    {

        //check the i'th char is a letter
        if (isalpha(plaintext[i]))
        {
            //check the i'th char is upper
            if (isupper(plaintext[i]))
            {
                printf("%c", ((plaintext[i] - 'A' + key[j]) % ALPHABET_SIZE) + 'A');
            }
            else
            {
                printf("%c", ((plaintext[i] - 'a' + key[j]) % ALPHABET_SIZE) + 'a');
            }

            //check whether incrementing j would exceed the bounds of the key array
            if (j + 1 > key_len -1)
            {
                //restart the cycle through the key array
                j = 0;
            }
            else
            {
                j++;
            }
        }
        else //i'th char is not a letter, just reprint it
        {
            printf("%c", plaintext[i]);
        }
    }
    printf("\n");
}

File: vigenere.py
import sys
import cs50

ALPHABET_SIZE = 26

#check we have the right number of arguments
if len(sys.argv) != 2:
    print("Error: you must provide the key as a command line argument.")
    exit(1)

#check key is alphanumeric
if str.isalpha(sys.argv[1]) == False:
    print("Error: key must be alphanumeric")
    exit(2)

#get plaintext
print("plaintext:", end="")
plaintext = cs50.get_string()

print("ciphertext:", end="")

#convert key to numbers
key = []
for c in sys.argv[1]:
    key.append(ord(c.upper()) - ord('A'))

j = 0
key_len = len(key)

#iterate through the plaintext string
for c in plaintext:

    if str.isalpha(c) == True:

        if str.isupper(c) == True:
            print( chr( ((ord(c) - ord('A') + key[j]) % ALPHABET_SIZE) + ord('A') ), end="")

        else:
            print( chr( ((ord(c) - ord('a') + key[j]) % ALPHABET_SIZE) + ord('a') ), end="")

        #increment j, which moves us along the key.  Reset to 0 if about to go out of bounds
        if j + 1 > key_len - 1:
            j = 0
        else:
            j += 1

    else:

        #c is not a letter, just reprint it
        print(c, end="")

print("")